compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -810,28 +873,161 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+                            hir_id,
+                            body: self.lower_body(|this| {
+                                (
+                                    &[],
+                                    hir::Expr {
+                                        hir_id: this.next_id(),
+                                        kind: hir::ExprKind::Path(hir::QPath::Resolved(
+                                            None,
+                                            this.arena.alloc(hir::Path {
+                                                res: Res::Def(
+                                                    DefKind::ConstParam,
+                                                    def_id,
+                                                ),
+                                                span,
+                                                segments: arena_vec![this; hir::PathSegment::from_ident(this.lower_ident(param.ident))],
+                                            }),
+                                        )),
+                                        span,
+                                    },
+                                )
+                            }),
+                        };
+
+                        hir::GenericArg::Const(hir::ConstArg { value: ct, span })
+                    }
+                }
+            })
+            .collect()
+    }
+
     /// Lowers a lifetime binder that defines `generic_params`, returning the corresponding HIR
     /// nodes. The returned list includes any "extra" lifetime parameters that were added by the
     /// name resolver owing to lifetime elision; this also populates the resolver's node-id->def-id
     /// map, so that later calls to `opt_node_id_to_def_id` that refer to these extra lifetime
     /// parameters will be successful.
-    #[instrument(level = "debug", skip(self))]
+    #[instrument(level = "debug", skip(self, in_binder))]
     #[inline]
-    fn lower_lifetime_binder(
+    fn lower_lifetime_binder<R>(
         &mut self,
         binder: NodeId,
         generic_params: &[GenericParam],
-    ) -> &'hir [hir::GenericParam<'hir>] {
-        let mut generic_params: Vec<_> = self.lower_generic_params_mut(generic_params).collect();
+        copy_generics: bool,
+        in_binder: impl FnOnce(&mut Self, &'hir [hir::GenericParam<'hir>]) -> R,
+    ) -> R {
+        let mut new_remapping = FxHashMap::default();
+
         let extra_lifetimes = self.resolver.take_extra_lifetime_params(binder);
         debug!(?extra_lifetimes);
-        generic_params.extend(extra_lifetimes.into_iter().filter_map(|(ident, node_id, res)| {
-            self.lifetime_res_to_generic_param(ident, node_id, res)
-        }));
-        let generic_params = self.arena.alloc_from_iter(generic_params);
-        debug!(?generic_params);
 
-        generic_params
+        let mut hir_extra_lifetimes = Vec::new();
+
+        let current_node_id_to_def_id = self.resolver.node_id_to_def_id.clone();
```

Comment:
Why do we need this clone?

Reply:
This was done in order to allow us to lower `for<'a>` twice in two different items. So the second lower can insert in the remapping.